<a href="https://colab.research.google.com/github/Latamila/Latamila/blob/main/rainForestforCredit.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [ ]:
from google.colab import files


In [ ]:
uploaded = files.upload()

Saving credito.csv to credito.csv


In [ ]:
credito = pd.read_csv('credito.csv')

In [ ]:
credito.head(10)

checking_balance  months_loan_duration credit_history               purpose  \
0           < 0 DM                     6       critical  furniture/appliances   
1       1 - 200 DM                    48           good  furniture/appliances   
2          unknown                    12       critical             education   
3           < 0 DM                    42           good  furniture/appliances   
4           < 0 DM                    24           poor                   car   
5          unknown                    36           good             education   
6          unknown                    24           good  furniture/appliances   
7       1 - 200 DM                    36           good                   car   
8          unknown                    12           good  furniture/appliances   
9       1 - 200 DM                    30       critical                   car   

   amount savings_balance employment_duration  percent_of_income  \
0    1169         unknown           > 7 years                  4   
1    5951        < 100 DM         1 - 4 years                  2   
2    2096        < 100 DM         4 - 7 years                  2   
3    7882        < 100 DM         4 - 7 years                  2   
4    4870        < 100 DM         1 - 4 years                  3   
5    9055         unknown         1 - 4 years                  2   
6    2835   500 - 1000 DM           > 7 years                  3   
7    6948        < 100 DM         1 - 4 years                  2   
8    3059       > 1000 DM         4 - 7 years                  2   
9    5234        < 100 DM          unemployed                  4   

   years_at_residence  age other_credit housing  existing_loans_count  \
0                   4   67         none     own                     2   
1                   2   22         none     own                     1   
2                   3   49         none     own                     1   
3                   4   45         none   other                     1   
4                   4   53         none   other                     2   
5                   4   35         none   other                     1   
6                   4   53         none     own                     1   
7                   2   35         none    rent                     1   
8                   4   61         none     own                     1   
9                   2   28         none     own                     2   

          job  dependents phone default  
0     skilled           1   yes      no  
1     skilled           1    no     yes  
2   unskilled           2    no      no  
3     skilled           2    no      no  
4     skilled           2    no     yes  
5   unskilled           2   yes      no  
6     skilled           1    no      no  
7  management           1   yes      no  
8   unskilled           1    no      no  
9  management           1    no     yes

In [ ]:
credito.shape

(1000, 17)

In [ ]:
#substituindo os valores das classes por 0 e 1
credito['default'] = credito['default'].replace({"no":0, "yes":1})

#trocando o nome da variavel alvo
credito = credito.rename(columns = {'default' : 'classe'})

In [ ]:
credito = credito.rename(columns = {'default' : 'classe',
                                    'checking_balance': 'saldo_bancario',
                                    'months_loan_duration': 'meses_financiamento',
                                    'credit_history': 'historico_credito',
                                    'purpose':'finalidade',
                                    'amount': 'valor',
                                    'savings_balance': 'saldo_aplicacoes',
                                    'employment_duration':'duracao_trabalho',
                                    'percent_of_income':'percentual_renda',
                                    'years_at_residence':'anos_residencia',
                                    'age': 'idade',
                                    'other_credit': 'outros_creditos',
                                    'housing': 'situacao_residencial',
                                    'existing_loans_count': 'emprestimos_existentes',
                                    'job': 'trabalho',
                                    'dependents': 'dependentes',
                                    'phone':'telefone'})

In [ ]:
credito

saldo_bancario  meses_financiamento historico_credito  \
0           < 0 DM                    6          critical   
1       1 - 200 DM                   48              good   
2          unknown                   12          critical   
3           < 0 DM                   42              good   
4           < 0 DM                   24              poor   
..             ...                  ...               ...   
995        unknown                   12              good   
996         < 0 DM                   30              good   
997        unknown                   12              good   
998         < 0 DM                   45              good   
999     1 - 200 DM                   45          critical   

               finalidade  valor saldo_aplicacoes duracao_trabalho  \
0    furniture/appliances   1169          unknown        > 7 years   
1    furniture/appliances   5951         < 100 DM      1 - 4 years   
2               education   2096         < 100 DM      4 - 7 years   
3    furniture/appliances   7882         < 100 DM      4 - 7 years   
4                     car   4870         < 100 DM      1 - 4 years   
..                    ...    ...              ...              ...   
995  furniture/appliances   1736         < 100 DM      4 - 7 years   
996                   car   3857         < 100 DM      1 - 4 years   
997  furniture/appliances    804         < 100 DM        > 7 years   
998  furniture/appliances   1845         < 100 DM      1 - 4 years   
999                   car   4576     100 - 500 DM       unemployed   

     percentual_renda  anos_residencia  idade outros_creditos  \
0                   4                4     67            none   
1                   2                2     22            none   
2                   2                3     49            none   
3                   2                4     45            none   
4                   3                4     53            none   
..                ...              ...    ...             ...   
995                 3                4     31            none   
996                 4                4     40            none   
997                 4                4     38            none   
998                 4                4     23            none   
999                 3                4     27            none   

    situacao_residencial  emprestimos_existentes    trabalho  dependentes  \
0                    own                       2     skilled            1   
1                    own                       1     skilled            1   
2                    own                       1   unskilled            2   
3                  other                       1     skilled            2   
4                  other                       2     skilled            2   
..                   ...                     ...         ...          ...   
995                  own                       1   unskilled            1   
996                  own                       1  management            1   
997                  own                       1     skilled            1   
998                other                       1     skilled            1   
999                  own                       1     skilled            1   

    telefone  classe  
0        yes       0  
1         no       1  
2         no       0  
3         no       0  
4         no       1  
..       ...     ...  
995       no       0  
996      yes       0  
997       no       0  
998      yes       1  
999       no       0  

[1000 rows x 17 columns]

In [ ]:
credito['classe'].value_counts()

0    700
1    300
Name: classe, dtype: int64

In [ ]:
credito['historico_credito'].value_counts()


good         530
critical     293
poor          88
very good     49
perfect       40
Name: historico_credito, dtype: int64

Quando usamos o método pop(), estamos, literalmente, extraindo a 
coluna do dataframe. Assim, a coluna “classe”, a nossa variável alvo, será armazenada 
em uma variável chamada “classe” e será eliminada do dataframe “credito”.

In [ ]:
#extraindo as classes do dataframe

classes= np.array(credito.pop('classe'))

O algoritmo que utilizaremos só funciona com dados 
numéricos, e você deve ter percebido que temos muitos dados não numéricos no 
nosso conjunto. Dessa forma, não conseguiremos criar o modelo. O pacote Pandas nos oferece um método para converter dados não numéricos em dados numéricos, o método get_dummies(). O método usa uma técnica 
conhecida como one-hot encoding para criar colunas adicionais no dataframe com valores binários.


In [ ]:
credito = pd.get_dummies(credito)

In [ ]:
credito

meses_financiamento  valor  percentual_renda  anos_residencia  idade  \
0                      6   1169                 4                4     67   
1                     48   5951                 2                2     22   
2                     12   2096                 2                3     49   
3                     42   7882                 2                4     45   
4                     24   4870                 3                4     53   
..                   ...    ...               ...              ...    ...   
995                   12   1736                 3                4     31   
996                   30   3857                 4                4     40   
997                   12    804                 4                4     38   
998                   45   1845                 4                4     23   
999                   45   4576                 3                4     27   

     emprestimos_existentes  dependentes  saldo_bancario_1 - 200 DM  \
0                         2            1                          0   
1                         1            1                          1   
2                         1            2                          0   
3                         1            2                          0   
4                         2            2                          0   
..                      ...          ...                        ...   
995                       1            1                          0   
996                       1            1                          0   
997                       1            1                          0   
998                       1            1                          0   
999                       1            1                          1   

     saldo_bancario_< 0 DM  saldo_bancario_> 200 DM  ...  \
0                        1                        0  ...   
1                        0                        0  ...   
2                        0                        0  ...   
3                        1                        0  ...   
4                        1                        0  ...   
..                     ...                      ...  ...   
995                      0                        0  ...   
996                      1                        0  ...   
997                      0                        0  ...   
998                      1                        0  ...   
999                      0                        0  ...   

     outros_creditos_store  situacao_residencial_other  \
0                        0                           0   
1                        0                           0   
2                        0                           0   
3                        0                           1   
4                        0                           1   
..                     ...                         ...   
995                      0                           0   
996                      0                           0   
997                      0                           0   
998                      0                           1   
999                      0                           0   

     situacao_residencial_own  situacao_residencial_rent  trabalho_management  \
0                           1                          0                    0   
1                           1                          0                    0   
2                           1                          0                    0   
3                           0                          0                    0   
4                           0                          0                    0   
..                        ...                        ...                  ...   
995                         1                          0                    0   
996                         1                          0                    1   
997                         1                          0                    0   

In [ ]:
import sklearn

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
treino,teste,treino_classes,teste_classes = train_test_split(credito, classes,
                                                             stratify= classes,
                                                             test_size=0.2,
                                                             random_state=123)

In [ ]:
#carregando o pacote para o modelo de arvore de decisao 
from sklearn.tree import DecisionTreeClassifier

In [ ]:
arvore=DecisionTreeClassifier(random_state=123)

In [ ]:
arvore.fit(treino, treino_classes)

DecisionTreeClassifier(random_state=123)

In [ ]:
#previsoes
previsoes_teste = arvore.predict(teste)

In [ ]:
#carregando os pacotes para avaliacao do modelo
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [ ]:
#matriz de confusao
confusion_matrix(previsoes_teste, teste_classes)

array([[103,  31],
       [ 37,  29]])

A matriz de confusão (confusion_matrix) é uma das abordagens mais utilizadas 
para avaliar modelos de classificação. Na diagonal principal da matriz (103 e 
29), estão os acertos do modelo. Perceba que usamos a função confusion_matrix 
com os parâmetros: previsoes_teste e teste_classes, ou seja, as previsões feitas pelo 
modelo e as classes reais do conjunto de dados. Assim, é possível saber quando 
o modelo acertou ou errou em suas previsões. 

**Nesse caso específico, o modelo acertou 132 previsões (103 + 29)**

A diagonal 
secundária apresenta os erros do modelo: 68 (37 + 31). Podemos, assim, calcular a 
precisão do nosso modelo, simplesmente, dividindo os acertos pelo total de previsões: 132 / 200 = 0,66 (66%). A precisão total do modelo é de 66%. Se avaliarmos 
somente a precisão total, diríamos que temos um modelo promissor.

**a função 
classification_report nos fornece um relatório mais detalhado do desempenho 
do modelo**

In [ ]:
print(classification_report(teste_classes, previsoes_teste))

              precision    recall  f1-score   support

           0       0.77      0.74      0.75       140
           1       0.44      0.48      0.46        60

    accuracy                           0.66       200
   macro avg       0.60      0.61      0.61       200
weighted avg       0.67      0.66      0.66       200



Preste atenção nos valores em destaque! A primeira linha mostra informações 
sobre a classe 0. Você deve estar lembrado(a) de que substituímos os valores “no” 
(não) para 0 e os valores “yes” (sim) para 1. 

a classe 0 representa os pareceres negativos e, nesse caso, o modelo teve 77% de precisão quando previu que 
a classe seria 0. Já para a classe 1, a precisão foi de 44%. A seta vermelha destaca 
a precisão total do modelo.

O modelo até se comporta bem quando faz previsões para a classe 0, mas 
tem um desempenho muito ruim quando a classe é 1

O modelo é muito mais preciso para a classe 
0, porque “aprendeu” mais sobre essa classe. 

O algoritmo random 
forest é uma evolução da árvore de decisão. Em linhas gerais, ele trabalha com 
várias árvores de forma simultânea, fazendo combinações diferentes de dados 
e variáveis — random forest: floresta randômica (GRUS, 201). Essa abordagem, 
em muitos casos, consegue construir modelos mais precisos. Verifiquemos! 

In [ ]:
#carregando o pacote do classificador random forest
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rf= RandomForestClassifier(n_estimators=100,
                           random_state=123,
                           max_features='sqrt',
                           n_jobs=1,
                           verbose=1)

In [ ]:
rf.fit(treino,treino_classes)
previsoes_teste_rf = rf.predict(teste)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    1.1s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.0s finished


Não se preocupe com a quantidade de parâmetros na criação do modelo. Veja, 
porém, que temos um parâmetro chamado n_estimators = 100. Esse parâmetro 
define, basicamente, a quantidade de árvores que o modelo manipulará. Isso é 
muito mais eficiente do que trabalhar com apenas uma única árvore. Será mesmo? Avaliaremos o resultado das previsões com o novo modelo

In [ ]:
confusion_matrix(previsoes_teste_rf, teste_classes)

array([[124,  32],
       [ 16,  28]])

In [ ]:
print(classification_report(teste_classes, previsoes_teste_rf))

              precision    recall  f1-score   support

           0       0.79      0.89      0.84       140
           1       0.64      0.47      0.54        60

    accuracy                           0.76       200
   macro avg       0.72      0.68      0.69       200
weighted avg       0.75      0.76      0.75       200



**Veja que a precisão total do modelo aumentou para 76%. As precisões das classestambém melhoraram. Temos, agora, para a classe 0, uma precisão de 79% e, para a classe 1, uma precisão de 64%, ou seja, um aumento de 20 pontos percentuais 
em relação ao modelo anterior**

O que acabamos de fazer é uma estratégia comumente 
adotada, ou seja, construir modelos com abordagens diferentes usando o 
mesmo conjunto de dados. Durante a etapa de construção dos modelos, é provável que você construa várias versões, com o objetivo de melhorar os resultados.